CDAP-5183 Start zookeeper and kafka processes in Standalone if audit is enabled

gokulavasan opened the pull request:
JIRAs : https://issues.cask.co/browse/CDAP-5183
https://issues.cask.co/browse/CDAP-5475

Build : http://builds.cask.co/browse/CDAP-DUT3921

chtyim commented on cdap-kafka/pom.xml:
```diff
@@ -46,6 +46,36 @@
+    <plugins>
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-shade-plugin</artifactId>
```
Shouldn't shade. This will affect the packaging in distributed mode. We can just copy all modules and dependencies to a `target/libexec` directory so that cdap-standalone can pickup later (see cdap-spark-core pom.xml).

chtyim commented on cdap-common/src/main/resources/cdap-default.xml:
```diff
@@ -62,6 +62,14 @@
 
   <property>
+    <name>local.kafka.dir</name>
+    <value>libkafka</value>
```
Since this is only used by cdap-standalone, I suggest to put it in cdap-site.xml under the cdap-standalone source, otherwise, this option is visible for the cluster mode, which doesn't really make sense. Also, we kind of have all extension jars put under the `ext` directory (see security and core for example).

chtyim commented on cdap-common/src/main/java/co/cask/cdap/common/conf/Constants.java:
```diff
@@ -823,6 +823,8 @@
   }
 
   public static final String CFG_LOCAL_DATA_DIR = "local.data.dir";
+  public static final String CFG_LOCAL_ZOOKEEPER_SERVER_PORT = "local.zkserver.port";
```
These two are very standalone specific. Better put it inside `StandaloneMain`.

chtyim commented on cdap-standalone/pom.xml:
```diff
@@ -241,7 +253,7 @@
                   <overWriteIfNewer>true</overWriteIfNewer>
                   <excludeGroupIds>org.apache.hbase,asm</excludeGroupIds>
-                  <excludeArtifactIds>zkclient,servlet-api,cdap-cli,cdap-explore-jdbc</excludeArtifactIds>
+                  <excludeArtifactIds>zkclient,servlet-api,cdap-cli,cdap-explore-jdbc,cdap-kafka</excludeArtifactIds>
```
Make the dependency on `cdap-kafka` in provided scope, then you don't need to exclude in here.

chtyim commented on cdap-standalone/pom.xml:
```diff
@@ -265,6 +277,23 @@
+                <id>copy-kafka-server-artifact</id>
+                <phase>prepare-package</phase>
+                <goals>
+                  <goal>copy-dependencies</goal>
```
Instead of copy-dependencies, just use resource copy (again see how it copies cdap-spark-core).

chtyim commented on cdap-standalone/src/main/java/co/cask/cdap/StandaloneMain.java:
```diff
@@ -126,7 +140,7 @@ private StandaloneMain(List<Module> modules, CConfiguration cConf) {
 
     // Start ZK client and Kafka client only when audit is enabled
-    boolean auditEnabled = cConf.getBoolean(Constants.Audit.ENABLED);
+    auditEnabled = cConf.getBoolean(Constants.Audit.ENABLED);
```
Why need to convert it to field? We already have the `zkClient` and `kafkaClient` instances. We can just check for null in the start/stop methods.

chtyim commented on cdap-standalone/src/main/java/co/cask/cdap/StandaloneMain.java:
```diff
@@ -200,6 +214,32 @@ public void startUp() throws Exception {
+      zkProcessBuilder.redirectErrorStream(true);
+      zookeeperProcess = zkProcessBuilder.start();
+      zkIOThread = new Thread(new IOHandlerThread(zookeeperProcess.getInputStream(), System.out));
+      zkIOThread.start();
```
Better wrap this logic into a class like `ExternalProcessRunner`. We actually have a generic implementation in https://github.com/caskdata/cdap-apps/blob/develop/TwitterSentiment/src/main/java/co/cask/cdap/apps/flowlet/ExternalProgramExecutor.java

You can modify based on the need (e.g. making it a `ExternalJavaProcessRunner`, so that it constructs the bin/java, classpath,.. etc).

gokulavasan commented:
@chtyim Thanks for the comments! Addressed them all. Please review when you get a chance. Thanks!

chtyim commented on cdap-standalone/src/main/java/co/cask/cdap/ExternalJavaProcessExecutor.java:
```diff
@@ -0,0 +1,141 @@
+    this(className, null);
+  }
+
+  ExternalJavaProcessExecutor(String className, @Nullable List<String> args) {
```
Why allow null? In general should use empty collection instead of null.

chtyim commented on cdap-standalone/src/main/java/co/cask/cdap/ExternalJavaProcessExecutor.java:
```diff
@@ -0,0 +1,141 @@
+    return new Runnable() {
+      @Override
+      public void run() {
+        BufferedReader reader = new BufferedReader(new InputStreamReader(process.getInputStream(), Charsets.UTF_8));
```
Should use the `try (resource) { }` syntax.

chtyim commented:
Just two comments left.

gokulavasan commented:
@chtyim Thanks for the review! Addressed the remaining comments. Please review. I will squash commits and wait for build to pass and then merge. Thanks again!

chtyim commented:
LGTM
